waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,376 @@
+          {:keys [authorization/principal authorization/user] :as auth-params-map}
+          (auth/auth-params-map saml-principal)]
+      (auth/handle-request-auth (constantly {:body ""
+                                             :headers {"location" redirect-url}
+                                             :status 303})
+                                request principal auth-params-map password auth-cookie-age-in-seconds))
+    (throw (ex-info "Missing saml-auth-data from SAML authenticated redirect message"
+                    {:status 400}))))
+
+(defn- escape-xml-string
+  "Escape a string for use in an XML document."
+  [str]
+  (string/escape str {\' "&apos;"
+                      \" "&quot;"
+                      \& "&amp;"
+                      \< "&lt;"
+                      \> "&gt;"}))
+
+(let [saml-authentication-request-template-fn
+      (template/fn
+        [{:keys [time-issued saml-service-name saml-id acs-url idp-uri]}]
+        (slurp (io/resource "auth/saml-authentication-request.xml")))]
+  (defn render-saml-authentication-request-template
+    "Render the SAML authentication request XML."
+    [context]
+    (saml-authentication-request-template-fn (pc/map-vals escape-xml-string context))))
+
+(defn- create-request
+  "Return XML elements that represent a SAML 2.0 auth request."
+  [time-issued saml-service-name saml-id acs-url idp-uri]
+  (render-saml-authentication-request-template (utils/keys-map time-issued saml-service-name saml-id acs-url idp-uri)))
+
+(def instant-format (f/formatters :date-time-no-ms))
+
+(defn- make-issue-instant
+  "Converts a date-time to a SAML 2.0 time string."
+  []
+  (du/date-to-str (t/now) instant-format))
+
+(defn- create-request-factory!
+  "Creates new requests for a particular service, format, and acs-url."
+  [idp-uri saml-service-name acs-url]
+  ;;; Bootstrap opensaml when we create a request factory.
+  (DefaultBootstrap/bootstrap)
+  #(create-request (make-issue-instant)
+                   saml-service-name
+                   (str "WAITER-" (utils/unique-identifier))
+                   acs-url
+                   idp-uri))
+
+(defrecord SamlAuthenticator [auth-redirect-endpoint idp-uri password saml-request-factory saml-signature-validator]
+  auth/Authenticator
+  (wrap-auth-handler [_ request-handler]
+    (fn saml-authenticator-handler [{:keys [headers query-string request-method uri] :as request}]
+      (let [waiter-cookie (auth/get-auth-cookie-value (get headers "cookie"))
+            [auth-principal _ :as decoded-auth-cookie] (auth/decode-auth-cookie waiter-cookie password)]
+        (cond
+          ;; Use the cookie, if not expired
+          (auth/decoded-auth-valid? decoded-auth-cookie)
+          (let [auth-params-map (auth/auth-params-map auth-principal)
+                request-handler' (middleware/wrap-merge request-handler auth-params-map)]
+            (request-handler' request))
+          :else
+          (metrics/endpoint-with-waiter-metrics
```

Comment:
Do we need metrics for this function?

Reply:
metrics changed to 3 counters - acs endpoint, auth-redirect endpoint, and initial SAML auth handler